### PR TITLE
Implement receive batch methods

### PR DESCRIPTION
**Pull request by ShivangiReja**
This PR implements Receive method using pull model which supports both the receive modes:
   - Receive in PeekLock mode
   - Receive in ReceiveAndDelete mode

### CODE REVIEW

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpClient.cs [hunk 1]
```diff
@@ -521,31 +521,19 @@ public override TransportSender CreateSender(ServiceBusRetryPolicy retryPolicy)
...
+        /// <param name="receiveMode">The <see cref="ReceiveMode"/> used to specify how messages are received. Defaults to PeekLock mode.</param>
         /// <param name="prefetchCount">Controls the number of events received and queued locally without regard to whether an operation was requested.  If <c>null</c> a default will be used.</param>
         /// <param name="sessionId"></param>
         /// <param name="isSessionReceiver"></param>
         ///
         /// <returns>A <see cref="TransportConsumer" /> configured in the requested manner.</returns>
         ///
         public override TransportConsumer CreateConsumer(ServiceBusRetryPolicy retryPolicy,
-                                                         long? ownerLevel,
+                                                         ReceiveMode receiveMode,
```
Comment (JoshLove-msft) [inline sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpClient.cs]: This was added into the ServiceBusReceiverClientOptions.

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpConsumer.cs [hunk 2]
```diff
@@ -156,7 +154,7 @@ private void CloseLink(ReceivingAmqpLink link)
             var receivedMessageCount = 0;
             var failedAttemptCount = 0;
             var tryTimeout = _retryPolicy.CalculateTryTimeout(0);
-            var waitTime = (maximumWaitTime ?? tryTimeout);
+            var waitTime = tryTimeout;
             var link = default(ReceivingAmqpLink);
             var retryDelay = default(TimeSpan?);
             var amqpMessages = default(IEnumerable<AmqpMessage>);
```
Comment (JoshLove-msft) [inline sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpConsumer.cs]: Need to update to use the RetryPolicy.RunOperation approach.

File: sdk/servicebus/Azure.Messaging.ServiceBus/tests/ReceiverLiveTests.cs [hunk 3]
```diff
@@ -48,5 +47,52 @@ await foreach (ServiceBusMessage peekedMessage in peekedMessages)
...
+                {
+                    ReceiveMode = ReceiveMode.ReceiveAndDelete,
+                };
+                var receiver = new QueueReceiverClient(TestEnvironment.ServiceBusConnectionString, scope.QueueName, clientOptions);
+                var messageCount = 0;
+
+                await foreach (var item in receiver.ReceiveBatchAsync(10))
+                {
+                    messageCount++;
+                    Console.Write(item.Body + " ");
```
Comment (JoshLove-msft) [inline sdk/servicebus/Azure.Messaging.ServiceBus/tests/ReceiverLiveTests.cs]: Should remove

File: sdk/servicebus/Azure.Messaging.ServiceBus/tests/ReceiverLiveTests.cs [hunk 4]
```diff
@@ -48,5 +47,52 @@ await foreach (ServiceBusMessage peekedMessage in peekedMessages)
...
+                await sender.SendRangeAsync(GetMessages(10));
+
+                var clientOptions = new QueueReceiverClientOptions()
+                {
+                    ReceiveMode = ReceiveMode.ReceiveAndDelete,
+                };
+                var receiver = new QueueReceiverClient(TestEnvironment.ServiceBusConnectionString, scope.QueueName, clientOptions);
+                var messageCount = 0;
+
+                await foreach (var item in receiver.ReceiveBatchAsync(10))
```
Comment (JoshLove-msft) [inline sdk/servicebus/Azure.Messaging.ServiceBus/tests/ReceiverLiveTests.cs]: Should we try to receive more than we sent to verify that only our sent messages are received?
Or track the message Ids?

File: sdk/servicebus/Azure.Messaging.ServiceBus/tests/ReceiverLiveTests.cs [hunk 5]
```diff
@@ -48,5 +47,52 @@ await foreach (ServiceBusMessage peekedMessage in peekedMessages)
...
+
+                var receiver = new QueueReceiverClient(TestEnvironment.ServiceBusConnectionString, scope.QueueName);
+                var messageCount = 0;
+
+                await foreach (var item in receiver.ReceiveBatchAsync(10))
+                {
+                    messageCount++;
+                    Assert.AreEqual(item.SystemProperties.DeliveryCount, 1);
+                }
+                Assert.AreEqual(messageCount, 10);
```
Comment (JoshLove-msft) [inline sdk/servicebus/Azure.Messaging.ServiceBus/tests/ReceiverLiveTests.cs]: Should we add a peekRange to verify that the 10 messages are still there?

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpClient.cs (hunk 1 above)
Comment (JoshLove-msft) [reply sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpClient.cs]: Nevermind, I misread this.

File: sdk/servicebus/Azure.Messaging.ServiceBus/tests/ReceiverLiveTests.cs (hunk 3 above)
Comment (ShivangiReja) [reply sdk/servicebus/Azure.Messaging.ServiceBus/tests/ReceiverLiveTests.cs]: Removed!

File: sdk/servicebus/Azure.Messaging.ServiceBus/tests/ReceiverLiveTests.cs (hunk 4 above)
Comment (ShivangiReja) [reply sdk/servicebus/Azure.Messaging.ServiceBus/tests/ReceiverLiveTests.cs]: Updated the test cases to track message Ids.

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpConsumer.cs [hunk 6]
```diff
@@ -147,114 +147,86 @@ private void CloseLink(ReceivingAmqpLink link)
...
-                        else if (ex is AmqpException)
-                        {
-                            throw activeEx;
-                        }
-                        else
-                        {
-                            throw;
-                        }
-                    }
+                    //link.DisposeDelivery(message, true, AmqpConstants.AcceptedOutcome);
```
Comment (JoshLove-msft) [inline sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpConsumer.cs]: Do we need this commented out code when in ReceiveAndDelete mode? 
It looks like Track 1 does this.

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpConsumer.cs [hunk 7]
```diff
@@ -147,114 +147,86 @@ private void CloseLink(ReceivingAmqpLink link)
...
-            {
-                ServiceBusEventSource.Log.MessageReceiveError(EntityName, ex.Message);
-                throw;
-            }
-            finally
-            {
                 stopWatch.Stop();
-                ServiceBusEventSource.Log.MessageReceiveComplete(EntityName, receivedMessageCount);
+
+                receivedMessageCount = receivedMessages.Count;
```
Comment (JoshLove-msft) [inline sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpConsumer.cs]: Hmm can this be removed?

File: sdk/servicebus/Azure.Messaging.ServiceBus/tests/ReceiverLiveTests.cs [hunk 8]
```diff
@@ -54,17 +54,32 @@ public async Task ReceiveMessagesInPeekLockMode()
             await using (var scope = await ServiceBusScope.CreateWithQueue(enablePartitioning: false, enableSession: false))
             {
                 await using var sender = new QueueSenderClient(TestEnvironment.ServiceBusConnectionString, scope.QueueName);
-                await sender.SendRangeAsync(GetMessages(10));
+                IEnumerable<ServiceBusMessage> messages = GetMessages(10);
```
Comment (JoshLove-msft) [inline sdk/servicebus/Azure.Messaging.ServiceBus/tests/ReceiverLiveTests.cs]: nit make 10 a constant, or even add it as a param and add a few test cases.

**Comment by JoshLove-msft** [comment]
Should we add test cases for receiving with sessions?

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpConsumer.cs (hunk 6 above)
Comment (ShivangiReja) [reply sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpConsumer.cs]: Uncommitted this code!

Comment (ShivangiReja) [reply sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpConsumer.cs]: Uncommented this code!

**Comment by ShivangiReja** [comment]
>Should we add test cases for receiving with sessions?

I'll add session related test cases when I'll work on sessions.

**Review by JoshLove-msft — APPROVED** [review]

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpConsumer.cs (hunk 6 above)
Comment (ShivangiReja) [reply sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpConsumer.cs]: Uncommented this code; although I'm not sure how it gets used in ReceiveAndDelete.

**Pull request merged**